org.eclipse.m2e.pde/src/org/eclipse/m2e/pde/MavenTargetLocation.java

```diff
@@ -273,8 +285,67 @@ private boolean isPomType(Artifact artifact) {
 		return POM_PACKAGE_TYPE.equals(artifact.getExtension());
 	}
 
+	private File getFeatureFile(Artifact artifact, CacheManager cacheManager) {
+		File baseFile = artifact.getFile();
+		
+		if(baseFile == null) {
+			return null;
+		} else if (baseFile.isDirectory()) {
+			File featureFile = new File(baseFile, ICoreConstants.FEATURE_FILENAME_DESCRIPTOR);
+			return featureFile.exists() ? featureFile : null;
+		} else if (DEFAULT_PACKAGE_TYPE.equals(FilenameUtils.getExtension(baseFile.getName()))) {
+			File featureFile = unpackFeatureFile(artifact, cacheManager);
+			return featureFile;
+		}
+		
+		return null;
+	}
+	
+	private File unpackFeatureFile(Artifact artifact, CacheManager cacheManager) {
+		try {
+			return cacheManager.accessArtifactFile(artifact, (file) -> {
+				// Unpack feature.xml into the same directory as the jar
+				File featureFile = new File(file.getParentFile(), ICoreConstants.FEATURE_FILENAME_DESCRIPTOR);
+				
+				// May have already been unpacked -> reuse
+				if(featureFile.exists()) {
+					return featureFile;
+				}
+				
+				File markerFile = new File(file.getParentFile(), NOT_A_FEATURE);
+				
+				// Artifact has already been checked during an earlier cycle
+				if(markerFile.exists()) {
```

Comment:
Maybe you want to check that the lastChange() of the marker file is >= than the last change of the file :-)

Reply:
Yeah... I was only thinking about releases, where the artifact file wouldn't be updated. But for snapshots, it has to be done.
Updated.